### PR TITLE
[MIPS] Use softPromoteHalf legalization for fp16 rather than PromoteFloat

### PR DESCRIPTION
**Pull request by yingopq**
Fix part of #97975.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
> Fix #97975.

This only fixes it for the MIPS architecture, so should not close the issue with "Fix".

The fix itself here looks fine to me, but I feel like the test coverage is very weak. There's just that one test, and because it's not a generated one, it omits too much information to even tell if the assembly is correct.

File: llvm/lib/Target/Mips/MipsISelLowering.h [hunk 1]
```diff
@@ -390,8 +392,8 @@ class TargetRegisterClass;
           DAG.getLoad(Ty, DL, DAG.getEntryNode(), GOT,
                       MachinePointerInfo::getGOT(DAG.getMachineFunction()));
       unsigned LoFlag = IsN32OrN64 ? MipsII::MO_GOT_OFST : MipsII::MO_ABS_LO;
-      SDValue Lo = DAG.getNode(MipsISD::Lo, DL, Ty,
-                               getTargetNode(N, Ty, DAG, LoFlag));
+      SDValue Lo =
+          DAG.getNode(MipsISD::Lo, DL, Ty, getTargetNode(N, Ty, DAG, LoFlag));
```
[the remark this answers is not in the document]
Comment (yingopq) [reply llvm/lib/Target/Mips/MipsISelLowering.h:396]: I had used the `git clang-format ` tool before submitting this PR, and it automatically modified the irrelevant code format, so I submitted it together.

[the remark this answers is not in the document]
Comment (yingopq) [reply llvm/lib/Target/Mips/MipsISelLowering.h:396]: I would do it locally to check it again.

**Comment by yingopq** [comment]
> > Fix #97975.
> 
> This only fixes it for the MIPS architecture, so should not close the issue with "Fix".
> 
> The fix itself here looks fine to me, but I feel like the test coverage is very weak. There's just that one test, and because it's not a generated one, it omits too much information to even tell if the assembly is correct.

I would update the title and enhance the test file.

File: llvm/test/CodeGen/Mips/fp16-promote.ll
Comment (nikic) [inline llvm/test/CodeGen/Mips/fp16-promote.ll:1]: Can you please rebase over https://github.com/llvm/llvm-project/commit/9f81acf4ef39e05bb2833cb3d79914940f31ac6c?

Comment (yingopq) [reply llvm/test/CodeGen/Mips/fp16-promote.ll:1]: OK, updated.

Comment (yingopq) [reply llvm/test/CodeGen/Mips/fp16-promote.ll:1]: @nikic Please help review, thanks!

**Comment by yingopq** [comment]
Ping

**Comment by tgross35** [comment]
Is it possible to add a test with `to_half` and `from_half` from https://github.com/llvm/llvm-project/issues/97981? The codegen should show only `mov` operations with no calls to `__gnu_h2f_ieee` / `__gnu_f2h_ieee`

**Comment by nikic** [comment]
@tgross35 Done.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**